File: gto-support-db/src/main/java/org/ccci/gto/android/common/db/Table.java
package org.ccci.gto.android.common.db;

import android.os.Parcel;
import android.os.Parcelable;

import org.ccci.gto.android.common.db.Expression.Field;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

public final class Table<T> implements Parcelable {
    @NonNull
    final Class<T> mType;
    @Nullable
    private final String mAlias;

    @Nullable
    private transient String mSqlTable;
    @Nullable
    private transient String mSqlPrefix;

    private Table(@NonNull final Class<T> type, @Nullable final String alias) {
        mType = type;
        mAlias = alias;
    }

    @NonNull
    public static <T> Table<T> forClass(@NonNull final Class<T> type) {
        return new Table<>(type, null);
    }

    @NonNull
    public Table<T> as(@Nullable final String alias) {
        return new Table<>(mType, alias);
    }

    @NonNull
    public Field field(@NonNull final String field) {
        return new Field(this, field);
    }

    @SuppressWarnings("unchecked")
    Table(@NonNull final Parcel in, @Nullable final ClassLoader loader) {
        try {
            mType = (Class<T>) Class.forName(in.readString(), true, loader);
        } catch (final ClassNotFoundException e) {
            throw new RuntimeException(e);
        }
        mAlias = in.readString();
    }

    @NonNull
    String sqlTable(@NonNull final AbstractDao2 dao) {
        // build the name if we haven't built it already
        if (mSqlTable == null) {
            final StringBuilder sql = new StringBuilder(dao.getTable(mType));
            if (mAlias != null) {
                sql.append(" AS ").append(mAlias);
            }
            mSqlTable = sql.toString();
        }

        return mSqlTable;
    }

    @NonNull
    String sqlPrefix(@NonNull final AbstractDao2 dao) {
        if (mSqlPrefix == null) {
            mSqlPrefix = (mAlias != null ? mAlias : dao.getTable(mType)) + ".";
        }

        return mSqlPrefix;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(@NonNull final Parcel out, int flags) {
        out.writeString(mType.getName());
        out.writeString(mAlias);
    }

    public static final Creator<Table> CREATOR = new ClassLoaderCreator<Table>() {
        @Override
        public Table createFromParcel(@NonNull final Parcel in, @Nullable final ClassLoader loader) {
            return new Table(in, loader);
        }

        @Override
        public Table createFromParcel(@NonNull final Parcel in) {
            return new Table(in, null);
        }

        @Override
        public Table[] newArray(final int size) {
            return new Table[size];
        }
    };
}

File: gto-support-db/src/main/java/org/ccci/gto/android/common/db/Table.kt
package org.ccci.gto.android.common.db

import android.os.Parcelable
import androidx.annotation.RestrictTo
import kotlinx.parcelize.IgnoredOnParcel
import kotlinx.parcelize.Parcelize

@Parcelize
data class Table<T : Any> internal constructor(
    @RestrictTo(RestrictTo.Scope.LIBRARY_GROUP) val type: Class<T>,
    private val alias: String? = null,
) : Parcelable {
    companion object {
        @JvmStatic
        fun <T : Any> forClass(type: Class<T>) = Table(type)
        inline fun <reified T : Any> forClass() = forClass(T::class.java)
    }

    @Suppress("ktlint:standard:function-naming")
    fun `as`(alias: String?) = copy(alias = alias)
    fun field(field: String) = Expression.Field(this, field)

    fun <T2 : Any> join(target: Table<T2>) = Join.create<T, T2>(target)
    inline fun <reified T2 : Any> join() = join(forClass(T2::class.java))

    @Transient
    @IgnoredOnParcel
    private var sqlPrefix: String? = null
    @Transient
    @IgnoredOnParcel
    private var sqlTable: String? = null

    internal fun sqlPrefix(dao: AbstractDao) = sqlPrefix ?: "${alias ?: dao.tableName(type)}.".also { sqlPrefix = it }
    internal fun sqlTable(dao: AbstractDao) = sqlTable ?: buildString {
        append(dao.tableName(type))
        if (alias != null) append(" AS ").append(alias)
    }.also { sqlTable = it }
}
